src/cpp/session/modules/data/SessionData.cpp

```diff
@@ -25,22 +27,168 @@ using namespace rstudio::core ;
 
 namespace rstudio {
 namespace session {
-namespace modules { 
+namespace modules {
 namespace data {
 
+class AsyncDataPreviewRProcess : public async_r::AsyncRProcess
+{
+public:
+   static boost::shared_ptr<AsyncDataPreviewRProcess> create(
+           const json::JsonRpcRequest& request,
+           const json::JsonRpcFunctionContinuation& continuation)
+   {
+      boost::shared_ptr<AsyncDataPreviewRProcess> pDataPreview(
+                  new AsyncDataPreviewRProcess(request, continuation));
+      pDataPreview->start();
+      return pDataPreview;
+   }
+
+   void terminateProcess()
+   {
+      async_r::AsyncRProcess::terminate();
+   }
+
+private:
+   AsyncDataPreviewRProcess(
+           const json::JsonRpcRequest& request,
+           const json::JsonRpcFunctionContinuation& continuation) :
+       continuation_(continuation),
+       request_(request)
+   {
+   }
+
+   FilePath pathFromModulesSource(std::string sourceFile)
+   {
+      FilePath modulesPath = session::options().modulesRSourcePath();
+      FilePath srcPath = modulesPath.complete(sourceFile);
+
+      return srcPath;
+   }
+
+   FilePath pathFromSource(std::string sourceFile)
+   {
+      FilePath sourcesPath = session::options().coreRSourcePath();
+      FilePath srcPath = sourcesPath.complete(sourceFile);
+
+      return srcPath;
+   }
+
+   void start()
+   {
+      std::ostringstream jsonStream;
+      json::Value requestValue = request_.params;
+      json::write(requestValue, jsonStream);
+      std::string jsonData = jsonStream.str();
+
+      std::string cmd = ".rs.preview_data_import_async('" + jsonData + "')";
+
+      std::vector<core::FilePath> sources;
+      sources.push_back(pathFromSource("Tools.R"));
+      sources.push_back(pathFromModulesSource("ModuleTools.R"));
+      sources.push_back(pathFromModulesSource("SessionDataViewer.R"));
+      sources.push_back(pathFromModulesSource("SessionDataImportV2.R"));
+
+      async_r::AsyncRProcess::start(cmd.c_str(), FilePath(), async_r::R_PROCESS_VANILLA, sources);
+   }
+
+   void onCompleted(int exitStatus)
+   {
+      if (terminationRequested())
```

Comment:
You still have a callback waiting on the client though (the onError method of the ServerRequestCallback). I'd advocate for returning an explicit error state here, or even a success state which has a "cancelled" flag. This is because ServerRequestCallback is supposed to have a strong contract that onResponseReceived on onError will be called. By not satisfying the continuation we end up not calling one of those back and leaving a socket hanging. Perhaps you have some client logic that already deals with this but in any case I think it's important that on the server side we always satisfy continuations.